lapce-core/src/indent.rs

```diff
@@ -11,18 +11,19 @@ pub enum IndentStyle {
     Spaces(u8),
 }
 
+pub const DEFAULT_INDENT: IndentStyle = IndentStyle::Spaces(4);
+
 impl IndentStyle {
+    pub const LONGEST_INDENT: &'static str = "        "; // 8 spaces
+
     /// Creates an `IndentStyle` from an indentation string.
     ///
     /// For example, passing `"    "` (four spaces) will create `IndentStyle::Spaces(4)`.
     #[allow(clippy::should_implement_trait)]
     #[inline]
     pub fn from_str(indent: &str) -> Self {
-        // XXX: do we care about validating the input more than this?  Probably not...?
-        debug_assert!(!indent.is_empty() && indent.len() <= 8);
-
         if indent.starts_with(' ') {
-            IndentStyle::Spaces(indent.len() as u8)
+            IndentStyle::Spaces((indent.len() % Self::LONGEST_INDENT.len()) as u8)
```

Comment:
Why the remainder? I think I would prefer either the previous panicking, or a `min` call, or both.

Also, shouldn't this function be fallible, and check if the `indent` is a string of spaces? Or count the actual number of spaces? I don't know all the usages off the top of my head, but just looking at this function in isolation, the implementation has a lot of assumptions.

Reply:
> Why the remainder? I think I would prefer either the previous panicking, or a `min` call, or both.

I thought reminder is better than `debug_assert()` because it will not panic. But if assertion is preferred, I can revert this change.
 
> 
> Also, shouldn't this function be fallible, and check if the `indent` is a string of spaces? Or count the actual number of spaces? I don't know all the usages off the top of my head, but just looking at this function in isolation, the implementation has a lot of assumptions.

The original implementation did not do either of these so I followed suit. Perhaps we can make this function fallible in subsequent changes?

Reply:
I guess my main question was: "why pick remainder over anything else?". It will not panic, but as written, it is not correct and its overflow behaviour is surprising: if I call `from_str` with 8 spaces, the function returns `IndentStyle::Spaces(0)`. 9 spaces -> 1

I think saturating using `indent.len().min(...)` would work better. I think I would like to keep the panic, though, just so we can catch the intended-as-unreachable conditions.

> The original implementation did not do either

We can change a lot here to fix surprising/inconsistent/etc. behavior and we don't have to be quick about it because the code quality and its performance is not really critical. I leave it up to you whether to address things in this PR or later (or address them at all).